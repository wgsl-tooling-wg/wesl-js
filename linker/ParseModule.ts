import { srcLog, SrcMap } from "@wesl/mini-parse";
import {
  AbstractElem,
  AliasElem,
  ExportElem,
  FnElem,
  GlobalDirectiveElem,
  ModuleElem,
  StructElem,
  TreeImportElem,
  VarElem,
} from "./AbstractElems.ts";
import { parseWgslD } from "./ParseWgslD.ts";

/** module with exportable text fragments that are optionally transformed by a templating engine */
export interface TextModule {
  kind: "text";
  exports: TextExport[];
  fns: FnElem[];
  vars: VarElem[];
  structs: StructElem[];
  imports: TreeImportElem[];
  aliases: AliasElem[];
  globalDirectives: GlobalDirectiveElem[];

  modulePath: string; // full path to the module e.g "package/sub/foo", or "_root/sub/foo"

  /** original src for module */
  src: string;

  /** tracks changes through conditional processing for error reporting */
  srcMap: SrcMap;
}

/** an export elem annotated with the fn/struct to which it refers */
export interface TextExport extends ExportElem {
  ref: FnElem | StructElem;
}

export function parseModule(
  src: string,
  naturalModulePath: string,
  params: Record<string, any> = {},
): TextModule {
  const srcMap = new SrcMap(src);

  const parsed = parseWgslD(src, srcMap);
  const exports = findExports(parsed, srcMap);
  const fns = filterElems<FnElem>(parsed, "fn");
  const aliases = filterElems<AliasElem>(parsed, "alias");
  const globalDirectives = filterElems<GlobalDirectiveElem>(
    parsed,
    "globalDirective",
  );
  const imports = parsed.filter(
    e => e.kind === "treeImport",
  ) as TreeImportElem[];
  const structs = filterElems<StructElem>(parsed, "struct");
  const vars = filterElems<VarElem>(parsed, "var");
  const overridePath = filterElems<ModuleElem>(parsed, "module")[0]?.name;

  const modulePath = overridePath ?? naturalModulePath;
  // dlog({ modulePath, overridePath });
  const kind = "text";
  return {
    ...{ kind, src, srcMap, modulePath },
    ...{ exports, fns, structs, vars, imports },
    ...{ aliases, globalDirectives },
  };
}

export function filterElems<T extends AbstractElem>(
  parsed: AbstractElem[],
  kind: T["kind"],
): T[] {
  return parsed.filter(e => e.kind === kind) as T[];
}

function findExports(parsed: AbstractElem[], srcMap: SrcMap): TextExport[] {
  const results: TextExport[] = [];
  const exports = findKind<ExportElem>(parsed, "export");

  exports.forEach(([elem, i]) => {
    const next = parsed[i + 1];

    if (elem.kind === "export") {
      if (next?.kind === "fn" || next?.kind === "struct") {
        results.push({ ...elem, ref: next });
      } else {
        srcLog(srcMap, elem.start, `#export what? (#export a fn or struct)`);
      }
    }
  });
  return results;
}

function findKind<T extends AbstractElem>(
  parsed: AbstractElem[],
  kind: T["kind"],
): [T, number][] {
  return parsed.flatMap((elem, i) =>
    elem.kind === kind ? ([[elem, i]] as [T, number][]) : [],
  );
}
