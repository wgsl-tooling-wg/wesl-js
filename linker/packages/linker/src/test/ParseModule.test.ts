import { expect, test } from "vitest";
import { parseModule, TextModule } from "../ParseModule.js";

test("simple fn export", () => {
  const src = `
    export
    fn one() -> i32 {
      return 1;
    }
  `;
  const module = testParseModule(src);
  expect(module.exports.length).toBe(1);
  expect(module.exports).toMatchSnapshot();
});

test("simple fn import", () => {
  const src = `
    import bar/foo

    fn bar() { foo(); }
  `;
  const module = testParseModule(src);
  expect(module.imports.length).toBe(1);
  expect(module.imports).toMatchSnapshot();
});

test("read simple struct export", () => {
  const exportPrefix = `export`;
  const src = `
    struct Elem {
      sum: f32
    }
  `;
  const module = testParseModule(exportPrefix + "\n" + src);
  expect(module.exports.length).toBe(1);
  const firstExport = module.exports[0];
  expect(firstExport.ref.name).toBe("Elem");
});

test("read #module", () => {
  const src = `
    module my.module.com
    export fn foo() {}
  `;
  const textModule = testParseModule(src);
  expect(textModule.modulePath).toBe("my.module.com");
});

// test.skip("parse error shows correct line after @if", () => {});

test("import gleam style", () => {
  const src = `
    import my/foo

    fn bar() { foo(); }
  `;
  const module = testParseModule(src);
  expect(module.imports).toMatchSnapshot();
});

function testParseModule(src: string): TextModule {
  return parseModule(src, "./test.wgsl");
}
