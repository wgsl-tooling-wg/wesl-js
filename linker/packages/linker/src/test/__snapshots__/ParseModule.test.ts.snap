// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`import gleam style 1`] = `
[
  {
    "end": 19,
    "imports": ImportTree {
      "segments": [
        SimpleSegment {
          "args": undefined,
          "as": undefined,
          "name": "my",
        },
        SimpleSegment {
          "args": undefined,
          "as": undefined,
          "name": "foo",
        },
      ],
    },
    "kind": "treeImport",
    "start": 5,
  },
]
`;

exports[`simple fn export 1`] = `
[
  {
    "end": 12,
    "kind": "export",
    "ref": {
      "calls": [],
      "end": 55,
      "kind": "fn",
      "name": "one",
      "nameElem": {
        "end": 22,
        "kind": "fnName",
        "name": "one",
        "start": 19,
      },
      "start": 16,
      "typeRefs": [
        {
          "end": 31,
          "kind": "typeRef",
          "name": "i32",
          "start": 28,
        },
      ],
    },
    "start": 5,
  },
]
`;

exports[`simple fn import 1`] = `
[
  {
    "end": 20,
    "imports": ImportTree {
      "segments": [
        SimpleSegment {
          "args": undefined,
          "as": undefined,
          "name": "bar",
        },
        SimpleSegment {
          "args": undefined,
          "as": undefined,
          "name": "foo",
        },
      ],
    },
    "kind": "treeImport",
    "start": 5,
  },
]
`;
